nifi-nar-bundles/nifi-metrics-reporting-bundle/nifi-metrics-reporting-task/src/main/java/org/apache/nifi/metrics/reporting/reporter/service/GraphiteMetricReporterService.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.controller.AbstractControllerService;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.metrics.reporting.task.MetricsReportingTask;
+import org.apache.nifi.processor.util.StandardValidators;
+
+import javax.net.SocketFactory;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A controller service that provides metric reporters for graphite, can be used by {@link MetricsReportingTask}.
+ *
+ * @author Omer Hadari
+ */
+@Tags({"metrics", "reporting", "graphite"})
+@CapabilityDescription("A controller service that provides metric reporters for graphite. " +
+        "Used by MetricsReportingTask.")
+public class GraphiteMetricReporterService extends AbstractControllerService implements MetricReporterService {
+
+    /** Points to the hostname of the graphite listener. */
+    public static final PropertyDescriptor HOST = new PropertyDescriptor.Builder()
+            .name("host")
+            .displayName("host")
+            .description("The hostname of the carbon listener")
+            .required(true)
+            .addValidator(StandardValidators.URI_VALIDATOR)
+            .build();
+
+    /** Points to the port on which the graphite server listens. */
+    public static final PropertyDescriptor PORT = new PropertyDescriptor.Builder()
+            .name("port")
+            .displayName("port")
+            .description("The port on which carbon listens")
+            .required(true)
+            .addValidator(StandardValidators.PORT_VALIDATOR)
+            .build();
+
+    /** Points to the charset name that the graphite server expects. */
+    public static final PropertyDescriptor CHARSET = new PropertyDescriptor.Builder()
+            .name("charset")
+            .displayName("charset")
+            .description("The charset used by the graphite server")
+            .required(true)
+            .defaultValue("UTF-8")
+            .addValidator(StandardValidators.CHARACTER_SET_VALIDATOR)
+            .build();
+
+    /** List of property descriptors used by the service. */
+    private static final List<PropertyDescriptor> properties;
+
+    static {
+        final List<PropertyDescriptor> props = new ArrayList<>();
+        props.add(HOST);
+        props.add(PORT);
+        properties = Collections.unmodifiableList(props);
+    }
+
+    /** Graphite sender, a connection to the server. */
+    private GraphiteSender graphiteSender;
+
+    /**
+     * Create the {@link #graphiteSender} according to configuration.
+     *
+     * @param context used to access properties.
+     */
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) {
+        String host = context.getProperty(HOST).getValue();
+        int port = context.getProperty(PORT).asInteger();
```

Comment:
It would be good to support expression language for host and port so that people can easily use variables to change environments. 

You can just add `supportsExpressionLanguage(true)` on the property descriptors and then here when you get the values do `context.getProperty(HOST).evaluateAttributeExpressions().getValue()`